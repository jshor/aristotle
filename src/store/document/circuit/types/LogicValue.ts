const LogicValue = {
  TRUE: 1,
  FALSE: -1,
  UNKNOWN: 0
}

export default LogicValue
