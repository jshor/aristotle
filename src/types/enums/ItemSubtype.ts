enum ItemSubtype {
  None = 'None',

  CustomCircuit = 'CustomCircuit',

  /* input nodes */
  Switch = 'Switch',
  PushButton = 'PushButton',
  Clock = 'Clock',

  /* output nodes */
  Lightbulb = 'Lightbulb',
  DigitDisplay = 'DigitDisplay',

  /* buffers */
  Buffer = 'Buffer',
  Not = 'Not',

  /* tri-state */
  Tristate = 'Tristate',

  /* logic gates */
  And = 'And',
  Or = 'Or',
  Nand = 'Nand',
  Nor = 'Nor',
  Xnor = 'Xnor',
  Xor = 'Xor'
}

export default ItemSubtype
