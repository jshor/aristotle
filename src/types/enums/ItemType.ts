enum ItemType {
  Buffer = 'Buffer',
  CircuitNode = 'CircuitNode',
  InputNode = 'InputNode',
  IntegratedCircuit = 'IntegratedCircuit',
  LogicGate = 'LogicGate',
  OutputNode = 'OutputNode'
}

export default ItemType
