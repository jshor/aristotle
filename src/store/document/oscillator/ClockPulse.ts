import LogicValue from '@/types/enums/LogicValue'
import Pulse from '@/types/interfaces/Pulse'
import { TinyEmitter } from 'tiny-emitter'
import { v4 as uuid } from 'uuid'

/**
 * This provides an oscillating clock wave.
 * It oscillates between {@link LogicValue.TRUE} and {@link LogicValue.FALSE}.
 */
export default class ClockPulse implements Pulse {
  /** Wave ID. */
  public id: string = uuid()

  /** User-friendly label for this wave. */
  public name: string

  /** Current signal value. */
  public currentValue: LogicValue = LogicValue.UNKNOWN

  /** The default signal value. */
  public defaultValue: LogicValue = LogicValue.UNKNOWN

  /** Interval, in milliseconds, to oscillate at. */
  public interval: number = 0

  /** Last update time, in milliseconds. */
  private lastUpdate: number

  /** Event emitter. */
  private emitter: TinyEmitter = new TinyEmitter()

  /** Whether or not the pulse is stopped. */
  private isStopped = false

  /**
   * Constructor.
   *
   * @param name - user-friendly label for this wave
   * @param interval - time, in milliseconds, to oscillate at
   * @param signal - current signal value to initialize at
   */
  constructor (name: string, interval: number, currentValue: LogicValue, defaultValue: LogicValue) {
    this.name = name
    this.currentValue = currentValue
    this.defaultValue = defaultValue
    this.interval = interval
    this.lastUpdate = -interval
  }

  /**
   * Deserializes the given data into a ClockPulse instance.
   */
  static deserialize (data?: {
    name: string
    interval: number
    currentValue: LogicValue
    defaultValue: LogicValue
  } | null) {
    if (data instanceof ClockPulse) return data
    if (!data) return

    return new ClockPulse(data.name, data.interval, data.currentValue, data.defaultValue)
  }

  /**
   * Returns an object representation of the ClockPulse instance.
   */
  toString = () => {
    return {
      name: this.name,
      interval: this.interval,
      currentValue: this.currentValue,
      defaultValue: this.defaultValue
    }
  }

  /**
   * Resets the current value to the default value.
   */
  reset = () => {
    this.lastUpdate = -this.interval
    this.currentValue = this.defaultValue
  }

  /**
   * Event listener.
   *
   * @param event - available values: `change`
   * @param fn - callback function, taking the oscillogram data as its sole argument
   */
  on = (event: string, fn: (signal: number) => void) => {
    this.emitter.on(event, fn)
  }

  /**
   * Stops the clock pulse.
   */
  stop = () => {
    this.isStopped = true
  }

  /**
   * Starts the clock pulse.
   */
  start = () => {
    this.isStopped = false
  }

  /**
   * Updates all waves with the given time elapsed.
   *
   * @param elapsed - time elapsed during simulation
   */
  public update = (elapsed: number): void => {
    if (!this.isStopped && elapsed >= this.lastUpdate + this.interval) {
      this.currentValue = this.currentValue === LogicValue.FALSE
        ? LogicValue.TRUE
        : LogicValue.FALSE
      this.lastUpdate = elapsed
      this.emitter.emit('change', this.currentValue)
    }
  }
}
