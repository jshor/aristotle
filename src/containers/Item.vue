<template>
  <draggable
    :position="position"
    :zoom="store.zoomLevel"
    :style="{
      transform: `rotate(${90 * rotation}deg)`,
      outline: 'none',
      zIndex
    }"
    :snap-mode="type === ItemType.Freeport ? 'radius' : 'outer'"
    :snap-boundaries="store.snapBoundaries"
    :bounding-box="boundingBox"
    :force-dragging="forceDragging"
    :aria-label="name"
    :title="name"
    :tabindex="0"
    @drag-start="onDragStart"
    @drag="onDrag"
    @drag-end="onDragEnd"
    @focus="onFocus"
    @blur="onBlur"
    @keydown.esc="onEscapeKey"
    @mousedown.stop="onMouseDown"
    ref="selectable"
    data-test="selectable"
  >
    <properties
      v-if="isPropertiesEnabled"
      :tabindex="0"
      :properties="properties"
      :id="id"
      @update="store.setProperties"
      @focus="onPropertiesFocus"
      @blur="onBlur"
      aria-label="Properties dialog"
      data-test="properties"
    />
    <selectable
      :is-selected="isSelected"
      :flash="flash"
    >
      <freeport v-if="type === ItemType.Freeport" />
      <integrated-circuit
        v-else-if="type === ItemType.IntegratedCircuit"
        :port-list="portList"
        @dblclick="$emit('openIntegratedCircuit')"
      />
      <input-switch
        v-else-if="type === ItemType.InputNode"
        :value="ports[0]?.value"
        @toggle="value => store.setPortValue({ id: ports[0].id, value })"
      />
      <lightbulb
        v-else-if="type === ItemType.OutputNode"
        :value="ports[0]?.value"
        :type="subtype"
      />
      <logic-gate
        v-else
        :input-count="(properties.inputCount?.value as number)"
        :type="subtype"
      />
      <port-set>
        <template
          v-for="(ports, orientation) in portList"
          v-slot:[orientation]
        >
          <port-item
            v-for="port in ports"
            :tabindex="0"
            :store="rawStore"
            :id="port.id"
            :key="port.id"
            :type="port.type"
            :is-freeport="port.isFreeport"
            :position="port.position"
            :orientation="port.orientation + rotation"
            :connected-port-ids="port.connectedPortIds"
            :rotation="rotation"
            @keypress.esc="onEscapeKey"
            @focus="onPortFocus(port.id)"
            @blur="$event => onPortBlur(port.id, $event)"
            @deselect="focus"
            data-test="port-item"
          />
        </template>
      </port-set>
    </selectable>
  </draggable>
</template>

<script lang="ts">
import ResizeObserver from 'resize-observer-polyfill'
import { StoreDefinition } from 'pinia'
import { defineComponent, PropType, onMounted, ref, ComponentPublicInstance, computed } from 'vue'
import Draggable from '@/components/Draggable.vue'
import LogicGate from '@/components/item/elements/LogicGate.vue'
import InputSwitch from '@/components/item/elements/InputSwitch.vue'
import IntegratedCircuit from '@/components/item/elements/IntegratedCircuit.vue'
import Lightbulb from '@/components/item/elements/Lightbulb.vue'
import Freeport from '@/components/item/Freeport.vue'
import PortHandle from '@/components/PortHandle.vue'
import PortSet from '@/components/item/PortSet.vue'
import Properties from '@/components/item/Properties.vue'
import Selectable from '@/components/Selectable.vue'
import PortItem from './PortItem.vue'
import ItemType from '@/types/enums/ItemType'
import ItemSubtype from '@/types/enums/ItemSubtype'
import DocumentState from '@/store/DocumentState'

/**
 * This component represents a two-dimensional item in the editor.
 * This can take on the form of a logic gate, input switch, lightbulb, user label, and more.
 * Items also contain ports, which are the circles that wires attach to to provide or send signals.
 */
export default defineComponent({
  name: 'Item',
  components: {
    Draggable,
    LogicGate,
    Freeport,
    InputSwitch,
    IntegratedCircuit,
    Lightbulb,
    PortHandle,
    PortSet,
    PortItem,
    Properties,
    Selectable
  },
  props: {
    /**
     * Bounding box of the item.
     */
    boundingBox: {
      type: Object as PropType<BoundingBox>,
      default: () => ({
        left: 0,
        top: 0,
        right: 0,
        bottom: 0
      })
    },

    /**
     * The directional rotation of the item.
     *
     * @values 0 = Left, 1, 2, 3
     */
    rotation: {
      type: Number,
      default: 0
    },

    /**
     * CSS z-index value.
     */
    zIndex: {
      type: Number,
      default: 1
    },

    /**
     * Whether or not the item is selected.
     */
    isSelected: {
      type: Boolean,
      default: false
    },

    /**
     * High-level taxonomy of the item.
     */
    type: {
      type: String as PropType<ItemType>,
      required: true
    },

    /**
     * More specific taxnommy of the item.
     */
    subtype: {
      type: String,
      required: true
    },

    /**
     * x-y position of the item in the editor.
     */
    position: {
      type: Object as PropType<Point>,
      required: true
    },

    /**
     * Optional user-friendly name of the item.
     */
    name: {
      type: String,
      default: 'Node'
    },

    /**
     * List of port IDs associated with the item.
     */
    portIds: {
      type: Array as PropType<string[]>,
      default: () => []
    },

    /**
     * Optional editable properties.
     */
    properties: {
      type: Object as PropType<PropertySet>,
      default: () => ({})
    },

    /**
     * ID of the group the item is associated to (if any).
     */
    groupId: {
      type: String,
      default: null
    },

    /**
     * Item ID.
     */
    id: {
      type: String,
      required: true
    },

    /** Whether or not the item should show a flash once to the user. */
    flash: {
      type: Boolean,
      default: false
    },

    /**
     * Document store instance.
     */
    store: {
      type: Function as PropType<StoreDefinition<string, DocumentState>>,
      required: true
    }
  },
  setup (props) {
    const store = props.store()
    const selectable = ref<ComponentPublicInstance<HTMLElement>>()
    const forceDragging = ref(false)

    let hasFocus = false
    let isDragging = false

    /* list of all ports that belong to this item */
    const ports = computed(() => {
      return props
        .portIds
        .map(portId => store.ports[portId])
    })

    /* slot-to-port-list map for displaying port-item elements */
    const portList = computed(() => {
      const locations = ['left', 'top', 'right', 'bottom']

      return ports.value.reduce((map: Record<string, Port[]>, port: Port) => ({
        ...map,
        [locations[port.orientation]]: [
          ...map[locations[port.orientation]],
          port
        ]
      }), locations.reduce((m, type) => ({
        ...m, [type]: []
      }), {
        left: [],
        top: [],
        right: [],
        bottom: []
      }))
    })

    /* whether or not the properties trigger button should be visible */
    const isPropertiesEnabled = computed(() => {
      return props.isSelected && hasFocus && Object.keys(props.properties).length > 0
    })

    /**
     * Places the window focus on the item element.
     */
    function focus () {
      selectable.value?.$el.focus()
      hasFocus = true
    }

    /**
     * Item drag start event handler.
     * This will compute the snappable boundaries on the editor.
     */
    function onDragStart () {
      store.setSnapBoundaries(props.id)
    }

    /**
     * Item drag event handler.
     * This will cache the undo-able state if dragged for the first time.
     *
     * @param position - new position that item has moved to
     */
    function onDrag (position: Point) {
      if (!isDragging) {
        isDragging = true
        store.cacheState() // cache the undo-able state in case a drag completes
      }

      store.setSelectionPosition({ id: props.id, position })
    }

    /**
     * Item drag end event handler.
     * This will commit the cached undo-able state.
     */
    function onDragEnd () {
      if (isDragging) {
        store.commitCachedState()
        isDragging = false
      }
    }

    /**
     * Inverts the selection state the item.
     */
    function onMouseDown ($event: MouseEvent) {
      if (props.isSelected) {
        store.deselectItem(props.id)
      } else {
        store.selectItem(props.id, $event.ctrlKey)
      }

      focus()
    }

    /**
     * Escape keydown event handler.
     * This will blur the element if neither it nor any of its children are focused.
     *
     * @param $event
     */
    function onEscapeKey ($event: KeyboardEvent) {
      if (document.activeElement === selectable.value?.$el) {
        selectable.value?.$el.blur()
      } else {
        $event.preventDefault()
      }
    }

    /**
     * Item blur event handler.
     * Updates `hasFocus` to false if the selection has left the item and any of its children.
     */
    function onBlur ($event: FocusEvent) {
      if (!selectable.value?.$el.contains($event.relatedTarget as Node)) {
        hasFocus = false
      }
    }

    /**
     * Item focus event handler.
     */
    function onFocus () {
      setTimeout(() => {
        if (!props.isSelected) {
          store.selectItem(props.id)
        }
      })

      hasFocus = true
    }

    /**
     * Properties dialog trigger button focus event handler.
     */
    function onPropertiesFocus () {
      store.setActivePortId(null)
      hasFocus = true
    }

    /**
     * Port focus event handler.
     *
     * @param portId - ID of the port focused
     */
    function onPortFocus (portId: string) {
      onFocus()
      store.setActivePortId(portId)
    }

    /**
     * Port blur event handler.
     *
     * @param portId - ID of the port blurred
     * @param $event - focus event
     */
    function onPortBlur (portId: string, $event: FocusEvent) {
      if (store.activePortId === portId) {
        store.setActivePortId(null)
      }
      onBlur($event)
    }

    /**
     * Item resize event handler. This will inform the store of the new size of the item.
     *
     * @param targets - target element that has been resized
     */
    function onSizeChanged ([ target ]: ResizeObserverEntry[]) {
      store.setItemSize({
        id: props.id,
        rect: target.contentRect
      })
    }

    const observer = new ResizeObserver(onSizeChanged)

    onMounted(() => {
      if (selectable.value) {
        observer.observe(selectable.value.$el)
      }

      if (store.activeFreeportId === props.id) {
        // if the item is an active freeport, then it is actively being drawn
        // allow for it to be moved when the mouse moves on first creation
        forceDragging.value = true
        store.activeFreeportId = null
        isDragging = true
      }

      store.setProperties({
        id: props.id,
        properties: props.properties
      })
    })

    return {
      selectable,
      store,
      ports,
      rawStore: props.store,
      portList,
      forceDragging,
      isPropertiesEnabled,
      focus,
      onDragStart,
      onDrag,
      onDragEnd,
      onMouseDown,
      onEscapeKey,
      onBlur,
      onFocus,
      onPropertiesFocus,
      onPortFocus,
      onPortBlur,
      onSizeChanged,
      ItemSubtype,
      ItemType
    }
  },
  watch: {
    isSelected (isSelected: boolean) {
      if (isSelected) {
        this.$el.focus()
      }
    }
  }
})
</script>
