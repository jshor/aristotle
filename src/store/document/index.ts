import { StoreGetters, defineStore } from 'pinia'
import { DocumentState, state } from './state'
import * as clipboardActions from './actions/clipboard'
import * as connectionActions from './actions/connection'
import * as editorActions from './actions/editor'
import * as groupingActions from './actions/grouping'
import * as itemActions from './actions/item'
import * as panningActions from './actions/panning'
import * as portActions from './actions/port'
import * as positioningActions from './actions/positioning'
import * as oscillatorActions from './actions/oscillator'
import * as rotationActions from './actions/rotation'
import * as selectionActions from './actions/selection'
import * as simulationActions from './actions/simulation'
import * as sizingActions from './actions/sizing'
import * as snappingActions from './actions/snapping'
import * as undoRedoActions from './actions/undoRedo'
import * as zIndexActions from './actions/zIndex'
import * as zoomingActions from './actions/zooming'
import * as getters from './getters'
import sortByZIndex from '@/utils/sortByZIndex'
import BaseItem from '@/types/interfaces/BaseItem'
import Item from '@/types/interfaces/Item'
import Connection from '@/types/interfaces/Connection'

export const createDocumentStore = (documentId: string) => defineStore({
  id: documentId,
  state,
  getters: {
    ...getters,

    baseItems (state) {
      return (Object
        .values(state.connections) as BaseItem[])
        .concat(Object.values(state.items) as BaseItem[])
        .sort(sortByZIndex)
    },

    hasSelection (state) {
      return state.selectedItemIds.size > 0 || state.selectedConnectionIds.size > 0
    },

    hasSelectedItems (state) {
      return state.selectedItemIds.size > 0
    },

    canGroup (state) {
      let ungroupedCount = 0
      let groupedCount = 0

      const hasUngrouped = (list: Set<string>, map: Record<string, Connection | Item>) => {
        for (const id of list) {
          map[id].groupId
            ? groupedCount++
            : ungroupedCount++

          if (ungroupedCount > 1 || (groupedCount && ungroupedCount)) {
            // at least two items are ungrouped, or at least one item is grouped and at least one item is ungrouped
            return true
          }
        }

        return false
      }

      return hasUngrouped(state.selectedItemIds, state.items) || hasUngrouped(state.selectedConnectionIds, state.connections)
    },

    canUngroup (state) {
      return state.selectedGroupIds.size > 0
    },

    canUndo (state) {
      return state.undoStack.length > 0
    },

    canRedo (state) {
      return state.redoStack.length > 0
    }
  },
  actions: {
    ...clipboardActions,
    ...connectionActions,
    ...editorActions,
    ...groupingActions,
    ...itemActions,
    ...panningActions,
    ...portActions,
    ...positioningActions,
    ...oscillatorActions,
    ...rotationActions,
    ...selectionActions,
    ...simulationActions,
    ...snappingActions,
    ...sizingActions,
    ...undoRedoActions,
    ...zIndexActions,
    ...zoomingActions
  }
})

export type DocumentStore = ReturnType<typeof createDocumentStore>

export type DocumentStoreInstance = ReturnType<DocumentStore>
