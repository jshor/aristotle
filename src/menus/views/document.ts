import { MenuFactory } from '@/types/interfaces/MenuFactory'
import { DocumentStore } from '@/store/document'
import { createHelpSubmenu } from '../submenus/help'
import { createEditSubmenu } from '../submenus/edit'
import { createFileSubmenu } from '../submenus/file'
import { createViewSubmenu } from '../submenus/view'
import { t } from '@/utils/i18n'

/**
 * Creates the document application menu.
 */
export const createDocumentMenu: MenuFactory = (useDocumentStore?: DocumentStore) => {
  return [
    {
      label: t('menu.file.parent'),
      submenu: createFileSubmenu(useDocumentStore)
    },
    {
      label: t('menu.edit.parent'),
      submenu: createEditSubmenu(useDocumentStore)
    },
    {
      label: t('menu.view.parent'),
      submenu: createViewSubmenu(useDocumentStore)
    },
    {
      label: t('menu.help.parent'),
      role: 'help',
      submenu: createHelpSubmenu()
    }
  ]
}
