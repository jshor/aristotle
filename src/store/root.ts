// @ts-check
import { defineStore } from 'pinia'
import { v4 as uuid } from 'uuid'
import basic from '../containers/fixtures/basic.json'
// import flipFlop from '../containers/fixtures/flipflop.json'
import flipFlop from '../containers/fixtures/counter.json'
// import integratedCircuit from '../containers/fixtures/ic.json'
import testIc from '../containers/fixtures/test.json'

import { createDocumentStore, DocumentStore } from './document'
import { useIntegratedCircuitStore } from './integratedCircuit'
import FileService from '@/services/FileService'
import getFileName from '@/utils/getFileName'
import idMapper from '@/utils/idMapper'
import Oscillogram from '@/types/types/Oscillogram'
import Item from '@/types/interfaces/Item'
import { DocumentStatus } from '@/types/enums/DocumentStatus'
import { ViewType } from '@/types/enums/ViewType'

export type RootStore = {
  documents: {
    [id: string]: {
      fileName: string
      filePath?: string
      displayName: string
      store: DocumentStore
    }
  }
  clipboard: string | null
  activeDocumentId: string | null
  canExit: boolean
  canPaste: boolean
  isDocumentSelectOpen: boolean
  isMobilePulloutOpen: boolean
  isFullscreen: boolean
  dialogType: ViewType
  isToolboxOpen: boolean
  navigationAnimationFrameId: number
  closeDialog?: () => Promise<boolean>
}

export const useRootStore = defineStore({
  id: 'root',
  state: (): RootStore => ({
    documents: {},
    clipboard: null,
    activeDocumentId: null,
    canExit: false,
    canPaste: false,
    isFullscreen: false,
    isToolboxOpen: false,
    isMobilePulloutOpen: false,
    dialogType: ViewType.None,
    isDocumentSelectOpen: false,
    navigationAnimationFrameId: 0
  }),
  getters: {
    hasOpenDocuments (state) {
      return Object
        .keys(state.documents)
        .length > 0
    },
    activeDocument (state) {
      if (state.activeDocumentId) {
        return state.documents[state.activeDocumentId]
      }

      return null
    },
    viewType (state) {
      if (state.dialogType) {
        return state.dialogType
      }

      return state.documents[state.activeDocumentId!]
        ? ViewType.Document
        : ViewType.None
    }
  },
  actions: {
    async launchIntegratedCircuitBuilder () {
      const integratedCircuitStore = useIntegratedCircuitStore()
      const document = this.activeDocument

      if (document) {
        await integratedCircuitStore.launchBuilder(document.store().$state)

        const store = document.store()

        // const dialogResult = window.api.showMessageBox({
        //   message: `Integrated circuit successfully created. Do you want to add it to the current document?`,
        //   title: 'Confirm',
        //   buttons: ['Yes', 'No']
        // })

        const idMappedIcItem = idMapper.mapIntegratedCircuitIds(integratedCircuitStore.model as Item)

        store.insertItemAtPosition({ item: idMappedIcItem })
        store.setItemBoundingBox(idMappedIcItem.id)
        store.setItemSelectionState(idMappedIcItem.id, true)
      }
    },
    openIntegratedCircuit (item: Item) {
      if (item.integratedCircuit) {
        this.openDocument('', item.integratedCircuit.serializedState, item.id, item.defaultName)
      }
    },
    async saveImage (data: Blob) {
      if (!this.activeDocument) return

      try {
        const partialName = this.activeDocument.displayName.replace(/\.[^/.]+$/, "")
        const fileName = `${partialName}.png`
        const filePath = window.api.showSaveFileDialog([
          {
            name: 'PNG Image',
            extensions: ['.png']
          }
        ], fileName)

        if (!filePath) return

        await FileService.writeBlob(filePath, data)
      } catch (error) {
        console.log('ERROR: ', error)
        window.api.showMessageBox({
          message: 'An error occurred while trying to save the image.',
          title: 'Error',
          type: 'error'
        })
      }
    },
    newDocument () {
      const id = uuid()
      const store = createDocumentStore(id)
      const document = store()

      document.initialize()

      this.documents[id] = {
        fileName: 'Untitled Circuit',
        displayName: 'Untitled Circuit',
        store
      }
      this.activateDocument(id)
    },

    async closeApplication (): Promise<boolean> {
      const dialogCloseResult = await this.closeDialog?.()

      if (dialogCloseResult === false) {
        return false
      }

      return this.closeAllDocuments()
    },

    /**
     * Attempts to close all open documents.
     *
     * @returns true if all documents were closed successfully, false otherwise
     */
    async closeAllDocuments (): Promise<boolean> {
      const openDocuments = [this.activeDocumentId!].concat(Object.keys(this.documents))

      for (let i = 0; i < openDocuments.length; i++) {
        if (this.documents[openDocuments[i]]) {
          const closedSuccessfully = await this.closeDocument(openDocuments[i])

          if (!closedSuccessfully) {
            return false
          }
        }
      }

      return true
    },

    /**
     * Closes the currently-active document.
     */
    async closeActiveDocument () {
      if (!this.activeDocumentId) return

      const closedSuccessfully = await this.closeDocument(this.activeDocumentId)

      if (closedSuccessfully) {
        await this.switchDocument(1)
      }
    },

    /**
     * Closes the document of the given ID.
     * This will prompt the user to save the document if it has unsaved changes.
     *
     * @returns true if the document was closed successfully, false otherwise
     */
    async closeDocument (id: string): Promise<boolean> {
      const document = this.documents[id]
      const store = document.store()

      if (!store.isDirty) {
        if (this.activeDocumentId === id) {
          this.switchDocument(1)
        }
        delete this.documents[id]
        return true
      }

      this.activateDocument(id)

      await store.load()

      const saveDialogResult = window
        .api
        .showMessageBox({
          message: `Save changes to ${document.displayName}?`,
          title: 'Confirm',
          buttons: ['Yes', 'No', 'Cancel']
        })

      if (saveDialogResult === 2) return false
      if (saveDialogResult === 0) {
        try {
          if (!this.attemptDocumentSave()) {
            return false
          }
        } catch (error) {
          const errorDialogResult = window
            .api
            .showMessageBox({
              message: 'An error occurred while trying to save the file. Close anyway?',
              title: 'Error',
              type: 'error',
              buttons: ['Yes', 'No']
            })

          if (errorDialogResult === 1) {
            return false
          }
        }
      }

      store.stopSimulation()
      store.$reset()
      store.$dispose()

      delete this.documents[id]

      return true
    },

    /**
     * Saves the document.
     *
     * @param saveAsNewFile - if true, forces saving the document as a new file
     */
    saveActiveDocument (saveAsNewFile: boolean = false) {
      try {
        this.attemptDocumentSave(saveAsNewFile)
      } catch (error) {
        window
          .api
          .showMessageBox({
            message: 'An error occurred while trying to save the file.',
            title: 'Error',
            type: 'error'
          })
      }
    },

    /**
     * Saves all open documents.
     */
    saveAllDocuments () {
      throw new Error('Not implemented.')
    },

    /**
     * Attempts to save the document as a file.
     * If the document has not been saved before, it will prompt the user to select a location to save the file.
     *
     * @param saveAsNewFile - if true, forces saving the document as a new file
     */
    attemptDocumentSave (saveAsNewFile: boolean = false) {
      if (!this.activeDocument) return true
      if (!this.activeDocument.filePath || saveAsNewFile) {
        // if no file name is defined yet, ask the user for the location to save it
        const fileName = this.activeDocument.fileName || `${this.activeDocument.displayName}.alfx`
        const filePath = window.api.showSaveFileDialog([
          {
            name: 'Aristotle Logic Circuit (*.alfx)',
            extensions: ['alfx']
          }
        ], fileName)

        if (filePath) {
          const fileName = getFileName(filePath)

          this.activeDocument.filePath = filePath
          this.activeDocument.fileName = fileName
          this.activeDocument.displayName = fileName
        }

        return false
      }

      const store = this.activeDocument.store()

      FileService.save(this.activeDocument.filePath!, {
        items: store.items,
        connections: store.connections,
        groups: store.groups,
        ports: store.ports
      })

      store.isDirty = false

      return true
    },
    async selectDocument () {
      const files = window.api.showOpenFileDialog([
        {
          name: 'Aristotle Logic Circuit (*.alfx)',
          extensions: ['alfx']
        },
        {
          name: 'Aristotle Integrated Circuit (*.aicx)',
          extensions: ['aicx']
        }
      ])

      for (let i = 0; i < files.length; i++) {
        await this.openDocumentFromPath(files[i])
      }
    },
    async openDocumentFromPath (filePath: string) {
      try {
        // TODO: need to create lock on file while open
        const content = await FileService.open(filePath)

        if (content) {
          this.openDocument(filePath, content, uuid())
        }
      } catch (error) {
        console.log(error)
        window.api.showMessageBox({
          message: `Could not open ${filePath}. The file is not a valid Aristotle circuit.`,
          title: 'Error',
          type: 'error'
        })
      }
    },
    openDocument (filePath: string, content: string, id: string, displayName: string = '') {
      if (this.documents[id]) {
        return this.activateDocument(id)
      }

      const store = createDocumentStore(id)
      const document = store()
      const fileName = getFileName(filePath)

      document.loadDocument(content)
      document.oscillator.onTick = (o: Oscillogram) => document.oscillogram = o

      this.documents[id] = {
        fileName,
        // if this is an integrated circuit file (.aicx), then it cannot be directly saved
        filePath: /^.*.aicx$/i.test(filePath)
          ? undefined // omit the file path for IC files
          : filePath, // otherwise, use the file path for a typical circuit
        displayName: displayName || fileName,
        store
      }
      this.activateDocument(id)
    },
    pauseActivity () {
      this.activeDocument?.store().stopSimulation()
    },
    resumeActivity () {
      this.activeDocument?.store().startSimulation()
    },
    activateDocument (id: string) {
      this.pauseActivity()
      this.activeDocumentId = id
      this.resumeActivity()
    },
    openTestDocuments () {
      this.openDocument('flip-flop.alfx', JSON.stringify(flipFlop), uuid(), 'flip-flop.alfx')
    },

    switchDocument (direction: number) {
      if (this.navigationAnimationFrameId) return

      this.navigationAnimationFrameId = requestAnimationFrame(async () => {
        const documentIds = Object.keys(this.documents)
        const currentIndex = documentIds.findIndex(id => id === this.activeDocumentId)
        let nextIndex = currentIndex + direction

        if (nextIndex < 0) nextIndex = documentIds.length - 1
        if (nextIndex >= documentIds.length) nextIndex = 0

        this.activateDocument(documentIds[nextIndex])
        await this.activeDocument?.store().load()
        this.navigationAnimationFrameId = 0
      })
    },

    toggleFullscreen () {
      this.isFullscreen = !this.isFullscreen

      window.api.setFullscreen(this.isFullscreen)
    },

    toggleToolbox () {
      this.isToolboxOpen = !this.isToolboxOpen
    },

    checkPasteability () {
      this.canPaste = window.api.canPaste()
    },

    setAllDocumentStatuses (status: DocumentStatus) {
      Object
        .keys(this.documents)
        .forEach(id => this.documents[id].store().setStatus(status))
    }
  }
})
