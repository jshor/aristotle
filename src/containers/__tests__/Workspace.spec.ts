import { mount, VueWrapper } from '@vue/test-utils'
import { setActivePinia, createPinia } from 'pinia'
import Item from '../Item.vue'
import { createDocumentStore, DocumentStore, DocumentStoreInstance } from '@/store/document'
import { createItem, createPort, stubAll } from '@/store/document/actions/__tests__/__helpers__'
import ItemType from '@/types/enums/ItemType'
import PortType from '@/types/enums/PortType'
import ItemSubtype from '@/types/enums/ItemSubtype'
import { useRootStore } from '@/store/root'
import type ItemInterface from '@/types/interfaces/Item'
import Port from '@/types/interfaces/Port'
import LogicValue from '@/types/enums/LogicValue'
import ClockPulse from '@/store/document/oscillator/ClockPulse'

setActivePinia(createPinia())

describe('Item container', () => {
  let wrapper: VueWrapper
  let storeDefinition: DocumentStore
  let store: DocumentStoreInstance
  let item1: ItemInterface
  let port1: Port, port2: Port

  const props = {
    zIndex: 1,
    isSelected: false,
    id: 'item1',
    flash: false
  }

  beforeEach(() => {
    storeDefinition = createDocumentStore('test-id')
    store = storeDefinition()

    item1 = createItem('item1', ItemType.InputNode, { portIds: ['port1', 'port2'] })
    port1 = createPort('port1', item1.id, PortType.Input, { isMonitored: true, hue: 12 })
    port2 = createPort('port2', item1.id, PortType.Output)

    store.ports = { port1, port2 }
    store.items = { item1 }
    store.setItemPosition({ id: item1.id, position: { x: 0, y: 0 } })

    wrapper = mount(Item, {
      props: {
        ...props,
        store: storeDefinition
      },
      global: {
        stubs: {
          Properties: true
        }
      }
    })
  })

  afterEach(() => {
    vi.resetAllMocks()
  })

  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should update the size when the number of input ports changes', async () => {
    stubAll(store, [
      'setItemSize',
      'setInputCount'
    ])

    const newValue = 3

    store.items.item1.properties = {
      inputCount: {
        value: 2,
        type: 'number',
        label: 'Input Count'
      }
    }

    await wrapper.vm.$nextTick()
    vi.resetAllMocks()

    store.items.item1.properties.inputCount!.value = newValue

    await wrapper.vm.$nextTick()

    expect(store.setInputCount).toHaveBeenCalledWith('item1', newValue)
    expect(store.setItemSize).toHaveBeenCalledWith({
      id: 'item1',
      rect: expect.any(Object)
    })
  })
})
