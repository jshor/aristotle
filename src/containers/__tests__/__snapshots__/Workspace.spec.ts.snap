// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Item container > should match the snapshot 1`] = `
"<div class="draggable" style="left: 0px; top: 0px; pointer-events: none; z-index: 1;" tabindex="0" data-test="item">
  <!--v-if-->
  <!--v-if-->
  <div class="circuit-component" style="transform: rotate(0deg);">
    <div class="switch switch--hi-z" type="InputNode" subtype="None" isselected="false" flash="false" name="" properties="[object Object]">
      <div class="switch__inner" tabindex="0" aria-role="switch">
        <div class="switch__trigger">
          <div class="switch__light"></div>
          <div class="switch__characters"></div>
        </div>
      </div>
    </div>
    <div class="circuit-component__ports--left circuit-component__ports">
      <div class="port-pivot" style="transform: rotate(-0deg);" data-port-id="port1">
        <div class="port-handle" id="port1" is-monitored="true" tabindex="0">
          <div class="port-handle__display">
            <div class="port-handle__hue"></div>
          </div>
        </div>
      </div>
      <div class="port-pivot" style="transform: rotate(-0deg);" data-port-id="port2">
        <div class="port-handle" id="port2" is-monitored="false" tabindex="0">
          <div class="port-handle__display">
            <div class="port-handle__hue"></div>
          </div>
        </div>
      </div>
    </div>
    <div class="circuit-component__ports--top circuit-component__ports"></div>
    <div class="circuit-component__ports--right circuit-component__ports"></div>
    <div class="circuit-component__ports--bottom circuit-component__ports"></div>
  </div>
</div>"
`;
